### DISCUSSION TITLE
Notebook as a `module` with `export`s

### DISCUSSION THREAD

**Discussion by greimel** [fonsp/Pluto.jl#2255] [Pluto development]
It would be nice if there was a way to specify that a notebook is wrapped into a `module`. Even better if one could specify `export`s. 

```julia
module FileName

export type1, type2, fun1, fun2

# notebook content

end
```

The default module name could be the filename without `.jl`.

Currently, when you want to develop a package in Pluto you need to create a second file, just to create the module and specify exports. See for example here

https://github.com/JuliaPluto/PlutoTest.jl/blob/9931f5a6a65cc8711cdf68829bf0f6a9d8093e1c/src/PlutoTest.jl#L1-L5

You can find such an instance even for some submodules of Pluto.jl itself.

https://github.com/fonsp/Pluto.jl/blob/72555eda2ba121a023243344ffa1dd75d3a17d35/src/webserver/Dynamic.jl#L80-L85

**Comment by disberd** [comment]
I personally don't think you would really get much benefit over just the one-liner with `module name include() end`, so in my opinion the addition and maintenance of code within Pluto to allow this is not justified.
If it's just for the convenience over the one-liner, I believe a macro doing that for you would be easier than modifying Pluto to have this behavior by default.

**Reply by greimel** [reply]
> I personally don't think you would really get much benefit over just the one-liner [...]

The problem is not so much that this is a one-liner, but a one-_filer_ ;-)

That is, if you cannot use a Pluto notebook as the single file of package. You must create a second file that defines the `module` and `include`s the notebook.

> [...] would be easier than modifying Pluto to have this behavior by default.

I wouldn't make this the default behaviour. I would like a GUI element in Pluto, where this can be switched on.

**Reply by disberd** [reply]
I see, I usually only work with modules as part of formal packages for which I still prefer to have a separate file for the Package definition and then include stuff there.
I only ever developed multi-notebook packages in Pluto so this problem does not apply there are you anyway need a separate file.

Can you share an example of usage of custom single package notebooks without a complete package structure with a Project.toml and other things? Just for curiosity of understanding the use-case.

**Reply by greimel** [reply]
Here's an example of a one-notebook package: https://github.com/JuliaPluto/PlutoTest.jl

I don't think it's possible to avoid the separate Project.toml. (But this would be another feature I'd like to have: export the Project.toml cell to the parent directory.)

**Reply by disberd** [reply]
Thanks, I knew about formal one-notebook packages like PlutoTest but I thought you had some different workflow where you were not making formal packages.
Once you have to create a separate folder for the Package including the notebook I don't see the presence of an additional file including the notebook as a big hurdle (especially since you only have to create it once and it stays the same even when you change the notebook). But that is just my opinion :D.
